File: 300-longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        int [][]dp = new int[n][n];
        for(int i=0;i<n;i++){
            Arrays.fill(dp[i] , -1);
        }
        return helper(0,-1,nums,dp);
    }
    private int helper(int idx,int prev,int []nums,int [][]dp){
        if(idx == nums.length){
            return 0;
        }
        if(dp[idx][prev+1] != -1){
            return dp[idx][prev+1];
        }
        int skip = helper(idx+1,prev,nums,dp);
        int take = 0;
        if(prev == -1 || nums[idx] > nums[prev]){
            take = 1 + helper(idx+1,idx,nums,dp);
        }
        int len = Math.max(skip,take);
        return dp[idx][prev+1] = len;
    }
}

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>>dp(n,vector<int>(n,-1));
        return solve(0,-1,nums,dp);
    }
    int solve(int idx,int prev,vector<int>& nums,vector<vector<int>>& dp){
        if(idx == nums.size()){
            return 0;
        }
        if(dp[idx][prev+1] != -1){
            return dp[idx][prev+1];
        }
        int skip = solve(idx+1,prev,nums,dp);
        int take = 0;
        if(prev == -1 || nums[idx] > nums[prev]){
            take = 1 +  solve(idx+1,idx,nums,dp);
        }
        int length = max(take,skip);
        return dp[idx][prev+1] = length;
    }
};